Enhanced support for Gerbil, Gambit

fare opened the pull request:
###### Motivation for this change

As Gerbil releases its v0.16, I have not only tweaked and cleaned up Gerbil support in Nix, but also added support for Gerbil libraries that may depend on each other, including one seed example, `gerbil-utils`.

Questions to reviewers:
- Am I doing the right thing when registering the packages in all-packages? Should the `gerbilPackages` be defined in their own file rather than in gerbil-support.nix ?
- For gerbil libraries, at the moment, I only intend to import "unstable" versions from git, to work with gerbil-unstable. Instead of `gerbilPackages.gerbil-utils-unstable`, should it be `gerbilPackage_unstable.gerbil-utils` or something?
- If I want to later provide some kind of wrapper in NixOS that export a `GERBIL_LOADPATH` and/or `NIX_GERBIL_LOADPATH` for all configured libraries, how do I do that? What about registering libraries for those in `~/.nix-profile` vs `/nix/var/nix/profiles` vs `/run/current-system/sw` ? Just `export GERBIL_LOADPATH=~/.nix-profile/gerbil/lib:/nix/var/nix/profiles/gerbil/lib:/run/current-system/sw/gerbil/lib` ? Should I create an option to enable this environment variable?
- Do I really need to put the version updates in a different commit than the `gerbil-utils` creation? Since the `gerbil-utils` creation is linked to all the nix code changes, is it OK if I put *that one* with them, and otherwise put the v0.16 update in its own *earlier* PR?

Notes:
- Gerbil stable v0.16 is not released quite yet, but will be in the next few days. Gerbil stable will depend on Gambit stable, Gerbil unstable will depend on Gambit unstable, even though in practice it will be the very same code, with only the version number difference.
- I will remove the [WIP] from this PR when Gerbil stable is out *and* reviewers have rubberstamped the general design, if not the actual code.
- At the same time, once the proper way to do it is agreed upon, I may add a dozen more gerbil libraries, by automating an import from the [gerbil directory](https://github.com/vyzo/gerbil-directory).

###### Things done

- Update `gambit-unstable`, `gerbil-unstable`. Create `gerbil-utils`.
- Move the Gambit files from `$out/` to `$out/gambit/` which makes the profile much cleaner
- Move the Gerbil files from `$out/` to `$out/gambit/` which makes the profile much cleaner
- Add support for Gerbil libraries, that may depend on other libraries (tested with an overlay).
- Create `gerbil-support.nix` for that.
- Return to `-O1` by default, but make that easily overridable.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

fare committed:
Refactor Gambit support

Refactor the build rule:
- Put files in $out/gambit instead of $out.
- Make the optimization setting easy to override.
- Make use of gccStdenv more explicit at this level.
- Support new-style runtime options for forcing UTF-8 I/O.
- Override the PACKAGE_VERSION and PACKAGE_STRING with git version.
- Note that the license is lgpl21, not lpgl2 (Note: also dual asl20).
- Try and fail to meaningfully add missing runtimeDeps.
- Build using NIX_BUILD_CORES.

fare committed:
Refactor Gerbil support

- Use the new Gambit support.
- Move files from $out to $out/gerbil.
- Use new Gerbil configuration and installation scripts.
- Move some fixups from preBuild to postPatch.
- Give up on previous failed attempts at using static libraries.
- Add support for compiling libraries written in Gerbil.
- Build using NIX_BUILD_CORES.
- Register all those things in all-packages.

fare commented:
@eraserhd @7c6f434c @thoughtpolice

7c6f434c commented on pkgs/development/compilers/gambit/bootstrap.nix:
```diff
@@ -1,6 +1,6 @@
-{ stdenv, fetchurl, autoconf, gcc, coreutils, ... }:
+{ gccStdenv, lib, fetchurl, autoconf, gcc, coreutils, ... }:
```
`gccStdenv` is for Darwin reasons, right?

fare replied:
I'm adding this comment:
```
# As explained in build.nix, GCC compiles Gambit 10x faster than Clang, for code 3x better
```

7c6f434c commented on pkgs/development/compilers/gambit/build.nix:
```diff
@@ -13,29 +15,35 @@
-  inherit version;
-  inherit src;
 
+  optimizationSetting = "-O1";
```
Wouldn't it be better in the `let` above `mkDerivation`

fare replied:
The ultimate goal is to make it configurable in an overlay. Should it be moved to `pkgs.gambit-support.optimizationSetting` or something?

7c6f434c replied:
Maybe make it an argument of the function then? Right now it is defined in a way that is very hard to override, as `configureFlags` use _Nix_ variable expansion, not built-time shell parameter expansion.

fare replied:
As in `gambitOptimizationSetting ? "-O1"` in the top parameters?

7c6f434c replied:
Yes.

7c6f434c commented on pkgs/development/compilers/gerbil/build.nix:
```diff
@@ -1,34 +1,51 @@
-{ stdenv, makeStaticLibraries,
+{ gccStdenv, lib, makeStaticLibraries, callPackage,
```
Is `callPackage` needed here?

fare replied:
Not anymore. I realize that makeStaticLibraries neither.

7c6f434c commented on pkgs/development/compilers/gerbil/build.nix:
```diff
@@ -1,34 +1,51 @@
   buildInputs_libraries = [ openssl zlib sqlite libxml2 libyaml libmysqlclient lmdb leveldb postgresql ];
-  buildInputs_staticLibraries = map makeStaticLibraries buildInputs_libraries;
+  # Use makeStaticLibraries to enable creation of statically linked binaries
+  #buildInputs_staticLibraries = map makeStaticLibraries buildInputs_libraries;
```
If it will no longer be needed, I guess it could be removed, and otherwise it's probably a good idea to write in the comment in which case it will come back.

7c6f434c commented on pkgs/development/compilers/gerbil/gerbil-tools.nix:
```diff
@@ -0,0 +1,53 @@
+let stdenv = gccStdenv; in # See ../gambit/build.nix
+
+rec {
+  gerbilPackage = {
```
Should the two `gerbilPackage` definitions be unified and just called with different arguments if needed?

fare replied:
Oops, I forgot to delete `gerbil-tools` after renaming it to `gerbil-support` (because there's a `gerbil-tools` package in the gerbil-directory).

fare commented:
PTAL. And please comment on the questions at top of PR.

7c6f434c commented on pkgs/development/compilers/gerbil/gerbil-support.nix:
```diff
@@ -0,0 +1,63 @@
+{ pkgs, gccStdenv, callPackage, fetchFromGitHub }:
+
+let stdenv = gccStdenv; in # See ../gambit/build.nix
```
Any reason not to use `gccStdenv` directly throughout this file?

7c6f434c commented on pkgs/top-level/all-packages.nix:
```diff
@@ -8131,10 +8131,13 @@ in
+  gerbil = callPackage ../development/compilers/gerbil { };
+  gerbil-unstable = callPackage ../development/compilers/gerbil/unstable.nix { };
+  gerbil-support = callPackage ../development/compilers/gerbil/gerbil-support.nix { };
+  gerbilPackage = gerbil-support.gerbilPackage;
```
Is there a reason to have this at the top level?

fare replied:
gerbilPackage at the top-level makes overlays slightly easier, but is just for convenience.

7c6f434c commented:
>    Am I doing the right thing when registering the packages in all-packages?

Mostly. I doubt it makes sense to give a top-level name to `gerbilPackage` function.

> Should the gerbilPackages be defined in their own file rather than in gerbil-support.nix ?

i think it doesn't matter at that size. With more packages new considerations might arise.

>    For gerbil libraries, at the moment, I only intend to import "unstable" versions from git, to work with gerbil-unstable. Instead of gerbilPackages.gerbil-utils-unstable, should it be gerbilPackage_unstable.gerbil-utils or something?

In principle, the first sounds like «unstable _library_ code» and the second like «stable library code running on unstable Gerbil». Not sure which is a closer description of the situation. 

Is there a plan for stable-Gerbil-using library packages in the foreseeable future?

 >   If I want to later provide some kind of wrapper in NixOS that export a GERBIL_LOADPATH and/or NIX_GERBIL_LOADPATH for all configured libraries, how do I do that? What about registering libraries for those in ~/.nix-profile vs /nix/var/nix/profiles vs /run/current-system/sw ? Just export GERBIL_LOADPATH=~/.nix-profile/gerbil/lib:/nix/var/nix/profiles/gerbil/lib:/run/current-system/sw/gerbil/lib ? Should I create an option to enable this environment variable?

Hmmm. I have a feeling that we generally claim that installing libraries into profiles does not work and it is not expected to work, one uses nix-shell (that runs `setup-hook`). So you probably need to define `setup-hook` that adds the Gerbil libraries from the build inputs to the environment variables (which is also needed for Gerbil packages to depend on each other). Then the recommended way of using libraries in an interactive shell would be nix-shell.

As someone who has stopped using NixOS mainline (and never used `nix-env -iA` much), I would prefer not to merge a NixOS change going against the recommended practices as I understand them.

>    Do I really need to put the version updates in a different commit than the gerbil-utils creation? Since the gerbil-utils creation is linked to all the nix code changes, is it OK if I put that one with them, and otherwise put the v0.16 update in its own earlier PR?

I think updating + creating the infrastructure in one commit, then adding the utils package in the next one is the closest to what people consider best practices here. In this specific case I don't care enough to actually enforce that it should be more than one commit.

fare commented:
Thanks a lot for the guidance.

> In principle, the first sounds like «unstable library code» and the second like «stable library code running on unstable Gerbil». Not sure which is a closer description of the situation.
> Is there a plan for stable-Gerbil-using library packages in the foreseeable future?

In the foreseeable future, all gerbil library packages will be unstable libraries running on top of unstable gerbil.

7c6f434c commented:
Is it a good idea to `recurseIntoAttrs` then? I would consider just having `gerbil-support-unstable` at the top level, that happens to include `gerbilPackages`

fare commented:
The point of `recurseIntoAttrs` is to make the gerbil libraries visible to `nix-env -qa`.

7c6f434c commented:
I understand that, but on one hand they are purely unstable and on the other hand `nix-env -qa` is slowly getting more and more recommended against as fantastically inefficient and on the third hand `nix-env -iA` of libraries is something that is widely expected not to work

7c6f434c commented on pkgs/development/compilers/gambit/build.nix:
```diff
@@ -1,41 +1,47 @@
-{ stdenv, git, openssl, autoconf, pkgs, makeStaticLibraries, version, gcc, src, coreutils }:
+{ gccStdenv, lib, git, openssl, autoconf, pkgs, makeStaticLibraries, gcc, coreutils, gnused, gnugrep,
+  src, version, git-version,
+  gambit-support, optimizationSetting ? "-O1", gambit-params ? pkgs.gambit-support.stable-params }:
```
Is `gambit-params` parameter just ignored in this file?

fare replied:
I had commented out the one use, pending debugging of other issues. It's back.

7c6f434c replied:
Right now Github shows me that it is still commented out. OK, I see it will be uncommented in the final version.

fare committed:
gambit-unstable: 2020-02-24 -> 2020-05-13

fare committed:
gerbil-unstable: 2020-02-27 -> 2020-05-17

fare committed:
gerbilPackages.gerbil-utils: init at 2020-05-16

fare commented:
OK. I am done tweaking the build.

I factored the package updates and the new package out of the build updates as such. I may update packages further (especially as a Gerbil release is looming) but I'm satisfied with this if you are.

The only thing is whether it's worth renaming gerbil-utils to gerbil-utils-unstable considering that there might not be a stable version for some time.

7c6f434c commented on pkgs/development/compilers/gambit/build.nix:
```diff
@@ -1,41 +1,47 @@
   buildInputs = [ git autoconf bootstrap openssl ];
 
+  # TODO: patch gambit's source so it has the full path to sed, grep, fgrep? Is there more?
+  runtimeDeps = [ gnused gnugrep ];
```
Does this actually do anything? I tried to find what would use this in Nixpkgs or Nix and found nothing…

7c6f434c reviewed (approved):
LGTM

fare committed:
gerbil: 0.15.1 -> 0.16

Now that v0.16 was released at last, make the configurePhase and instalPhase
the same again for default and unstable.

fare commented:
PTAL. Gerbil v0.16 was released, everything is now ready for final review and merge.
